rfcs/taskforce/0002-network-utility.md

```diff
@@ -0,0 +1,203 @@
+
+1. Keyset Generation (Account Creation)
+2. Genesis Ledger Creation
+3. Distribution of custom network configuration and keys
+4. Deployment of infrastructure with updated configuration
+5. Monitoring the health of the network and debuging issues / crashes
+
+This document proposes a tool that unifies these processes by collecting tools already written for these purposes and packaging them in a simple CLI and JS API.
+
+# Design
+
+## Existing System
+
+For these processes we mainly use the [testnet-keys.py](https://github.com/CodaProtocol/coda-automation/blob/master/scripts/testnet-keys.py) script in coda-automation.
+
+1. Keypair generation calls `coda advanced-generate keypair` generate new accounts
+2. Genesis ledger creation happens using custom logic inside this script
+3. Distribution of configuration uses Kubernetes secrets
+4. Deployment consists of a variety of terraform configurations targeting Google Kubernetes Engine
+5. We have a variety of charts and metrics for monitoring macro network health but debugging specific issues is often manual
+
+### Improvements
+
+While this system is effective, there are a variety of potential improvements to this process.
+
+- Keypair generation depends on the daemon
+- Genesis ledger logic could be cleaned up and made more flexible to support a wider variety of configuraitons
+- These tools could be packaged together with a JavaScript API for extending functionality
+- The internal data structures for storing these keypairs, genesis configuration and distribution are not well defined and could be more robust
+- Tools for deployment and monitoring are seperate and these could be packaged in one simple workflow that uses the same underlying functionality
+
+## Proposal
+
+All of these features should be included in a simple command line tool that also exposes a TypeScript API. This service will be written in Reason with TypeScript bindings.
+
+Since there may be some initial configuration required for the operation of a network, the following command will take care of setting up all the required services (google-cloud-sdk, aws, etc.)
+
+```
+coda-network init
+```
+
+### Keypair Management
+
+Keypairs have a private key and a corresponding public key. For new keypair generation, we can use the new client-sdk which is a much lighter dependency and has a clean JavaScript interface.
+
+However there are also existing keys which we'll need to be able to reference by public key alone since we don't control the private keys. An example of this would be community owned wallets for inclusion in the genesis ledger. Since this distinction can be made at a higher level, we'll keep the definition of a keypair to mean a public key for which we also have th private key.
+
+```
+coda-network keypair create
+```
+
+This command will generate a new keypair and save it locally.
+
+```
+coda-network keypair upload <PUBLIC_KEY>
```

Comment:
It's not clear to me why we would need keypair upload/download since we only care about upload/download of the keypairs when they are in keysets (this may make implementation easier as well). Is there a specific use case you had in mind? Also, if so, is that use case important for the "task force" world or could it be something we add later?

Reply:
I guess in my proposed universe the keyset would need to be smart enough to upload private keys associated with public keys if they exist.

Reply:
I've updated the keyset documentation to try to address this lmk if there's still ambiguity wrt to this. I think this will be more of a utility functionality that we'll need to expose on the keypair but we don't neccissarily need to expose it through the CLI.

Reply:
I think I understand -- makes sense to only expose it in the API probably. But does keyset need upload/download on the CLI (which then can call the individual key upload/download as necessary)